### PR TITLE
installer: handle installer failures better

### PR DESCRIPTION
**Pull request by sttts**
- do not proceed with other nodes if one node installation fails
- retry on OOM of the installer

### CODE REVIEW

File: pkg/operator/staticpod/controller/installer/installer_controller.go [hunk 1]
```diff
@@ -156,31 +157,65 @@ func nodeToStartRevisionWith(getStaticPodState func(nodeName string) (state stat
 		}
 	}
 
-	// otherwise try to find a node that is not ready
+	// otherwise try to find a node that is not ready. Take the oldest one.
+	oldestNotReadyRevisionNode := -1
+	oldestNotReadyRevision := math.MaxInt32
 	for i := range nodes {
```
Comment (mfojtik) [inline pkg/operator/staticpod/controller/installer/installer_controller.go]: nit: can you change `i` to `nodeID` ? makes the code more readable imho.

File: pkg/operator/staticpod/controller/installer/installer_controller.go [hunk 2]
```diff
@@ -156,31 +157,65 @@ func nodeToStartRevisionWith(getStaticPodState func(nodeName string) (state stat
 		}
 	}
 
-	// otherwise try to find a node that is not ready
+	// otherwise try to find a node that is not ready. Take the oldest one.
```
Comment (mfojtik) [inline pkg/operator/staticpod/controller/installer/installer_controller.go]: what does it mean "oldest" ? the one that has the lowest revision?

File: pkg/operator/staticpod/controller/installer/installer_controller.go [hunk 3]
```diff
@@ -156,31 +157,65 @@ func nodeToStartRevisionWith(getStaticPodState func(nodeName string) (state stat
...
+		}
+	}
+	if oldestNotReadyRevisionNode >= 0 {
+		return oldestNotReadyRevisionNode, nil
 	}
 
-	// last but not least, find a node that is has the wrong revision
+	// find a node that is has the wrong revision. Take the oldest one.
+	oldestPodRevisionNode := -1
+	oldestPodRevision := math.MaxInt32
```
Comment (mfojtik) [inline pkg/operator/staticpod/controller/installer/installer_controller.go]: we better not have more than 2147483647 revisions :D

File: pkg/operator/staticpod/controller/installer/installer_controller.go [hunk 4]
```diff
@@ -156,31 +157,65 @@ func nodeToStartRevisionWith(getStaticPodState func(nodeName string) (state stat
...
+			oldestNotReadyRevisionNode = i
+			oldestNotReadyRevision = revisionNum
+		}
+	}
+	if oldestNotReadyRevisionNode >= 0 {
+		return oldestNotReadyRevisionNode, nil
 	}
 
-	// last but not least, find a node that is has the wrong revision
+	// find a node that is has the wrong revision. Take the oldest one.
```
Comment (mfojtik) [inline pkg/operator/staticpod/controller/installer/installer_controller.go]: what means "wrong" revision? you looking for a node that is on the oldest revision?

**Comment by mfojtik** [comment]
/lgtm

We will add markdown explaining the logic of picking nodes and rolling revisions as a follow up.

File: pkg/operator/staticpod/controller/installer/installer_controller.go (hunk 3 above)
Comment (sttts) [reply pkg/operator/staticpod/controller/installer/installer_controller.go]: Not before OpenShift 5.0

**Pull request merged**